test/run.d

```diff
@@ -0,0 +1,308 @@
+    auto targets = args.predefinedTargets;
+    // normalize
+    targets = targets.map!(f => format!"%s/%s"(f.absolutePath.dirName.baseName, f.baseName)).array;
+    // filter
+    targets = targets.filterTargets;
+
+    if (targets.length > 0)
+    {
+        auto taskPool = new TaskPool(jobs);
+        scope(exit) taskPool.finish();
+        ensureToolsExists;
+        foreach (target; taskPool.parallel(targets, 1))
+        {
+            auto makeArguments = [scriptDir.buildPath("test_results", "d_do_test"), target];
+            spawnProcess(makeArguments, env, Config.none, scriptDir).wait;
+        }
+    }
+}
+
+/**
+Builds the binary of the tools required by the testsuite
+Does nothing if the tools already exist and are newer than their source
+*/
+void ensureToolsExists()
+{
+    static toolsDir = scriptDir.buildPath("tools");
+    resultsDir.mkdirRecurse;
+    auto tools = [
+        "d_do_test",
+        "sanitize_json",
+    ];
+    // TODO: run in parallel
+    foreach (tool; tools)
+    {
+        auto targetBin = resultsDir.buildPath(tool);
+        auto sourceFile = toolsDir.buildPath(tool ~ ".d");
+        if (targetBin.timeLastModified.ifThrown(SysTime.init) > sourceFile.timeLastModified)
+            writefln("%s is already up-to-date", tool);
+        else
+        {
+            auto command = ["dmd", "-of"~targetBin, sourceFile];
+            writefln("Executing: %-(%s %)", command);
+            spawnProcess(command).wait;
+        }
+    }
+
+    // ensure output directories exist
+    foreach (dir; ["compilable", "fail_compilation", "runnable"])
+        resultsDir.buildPath(dir).mkdirRecurse;
+}
+
+/**
+Goes through the target list and replaces short-hand targets with their expanded version.
+Special targets:
+- clean -> removes resultsDir + immediately stops the runner
+*/
+auto predefinedTargets(string[] targets)
+{
+    static findFiles(string dir)
+    {
+        return scriptDir.buildPath(dir).dirEntries("*{.d,.sh}", SpanMode.shallow).map!(e => e.name);
+    }
+
+    Appender!(string[]) newTargets;
+    foreach (t; targets)
+    {
+        switch (t)
+        {
+            case "clean":
+                resultsDir.rmdirRecurse;
+                exit(0);
+                break;
+
+            case "run_runnable_tests", "runnable":
+                newTargets.put(findFiles("runnable"));
+                break;
+
+            case "run_fail_compilation_tests", "fail_compilation", "fail":
+                newTargets.put(findFiles("fail_compilation"));
+                break;
+
+            case "run_compilable_tests", "compilable", "compile":
+                newTargets.put(findFiles("compilable"));
+                break;
+
+            case "all":
+                newTargets.put(findFiles("runnable"));
+                newTargets.put(findFiles("fail_compilation"));
+                newTargets.put(findFiles("compilable"));
+                break;
+
+            default:
+                newTargets ~= t;
```

Comment:
This results in an odd error in some cases.  Maybe the following check would work well?
```D
if (!testDirs.canFind(t.dirName))
{
    writefln("Error: invalid target '%s'", t);
    exit(1);
}
```

Reply:
We already check for file existence in `filterTargets`:

```
./run.d compilable/foo
Warning: compilable/foo can't be found
```

What's the odd error message you get?